src/Symfony/Component/DependencyInjection/SyntaxAware/SyntaxAwareDefinition.php

```diff
@@ -0,0 +1,173 @@
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function __construct($class = null, array $arguments = array())
+    {
+        $arguments = $this->resolveServices($arguments);
+
+        parent::__construct($class, $arguments);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setArguments(array $arguments)
+    {
+        $arguments = $this->resolveServices($arguments);
+
+        return parent::setArguments($arguments);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addArgument($argument)
+    {
+        $argument = $this->resolveServices($argument);
+
+        return parent::addArgument($argument);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replaceArgument($index, $argument)
+    {
+        $argument = $this->resolveServices($argument);
+
+        return parent::replaceArgument($index, $argument);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setProperties(array $properties)
+    {
+        $properties = $this->resolveServices($properties);
+
+        return parent::setProperties($properties);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setProperty($name, $value)
+    {
+        $value = $this->resolveServices($value);
+
+        return parent::setProperty($name, $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setFactory($factory)
+    {
+        if (is_string($factory)) {
+            if (strpos($factory, ':') !== false && strpos($factory, '::') === false) {
+                $parts = explode(':', $factory);
+                $factory = array($this->resolveServices('@'.$parts[0]), $parts[1]);
+            }
+        } else {
+            $factory = array($this->resolveServices($factory[0]), $factory[1]);
+        }
+
+        return parent::setFactory($factory);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setConfigurator($callable)
+    {
+        if (is_array($callable)) {
+            $callable = array($this->resolveServices($callable[0]), $callable[1]);
+        }
+
+        return parent::setConfigurator($callable);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addMethodCall($method, array $arguments = array())
+    {
+        $arguments = $this->resolveServices($arguments);
+
+        return parent::addMethodCall($method, $arguments);
+    }
+
+    /**
+     * Resolves services.
+     *
+     * @param string|array $value
+     *
+     * @return array|string|Reference
+     */
+    private function resolveServices($value)
+    {
+        if (is_array($value)) {
+            $value = array_map(array('self', 'resolveServices'), $value);
+        } elseif (is_string($value) &&  0 === strpos($value, '@=')) {
+            return new Expression(substr($value, 2));
+        } elseif (is_string($value) &&  0 === strpos($value, '@')) {
+            if (0 === strpos($value, '@@')) {
+                $value = substr($value, 1);
```

Comment:
I would use a return here directly instead of doing this weird logic with ``$invalidBehavior = null`` to detect that it should not be a reference